oauth2_lib/graphql_authentication.py

```diff
@@ -0,0 +1,128 @@
+
+logger = structlog.get_logger(__name__)
+
+
+class OauthContext(BaseContext):
+    def __init__(self, get_current_user: Callable[[], OIDCUserModel], get_opa_decision: Callable[[str], bool]):
+        self.get_current_user = get_current_user
+        self.get_opa_decision = get_opa_decision
+        super().__init__()
+
+
+OauthInfo = Info[OauthContext, RootValueType]
+
+
+def get_path_as_string(path: Path) -> str:
+    if path.prev:
+        return f"{get_path_as_string(path.prev)}/{path.key}"
+    else:
+        return f"{path.key}"
+
+
+class IsAuthenticated(BasePermission):
+    message = "User is not authenticated"
+
+    async def has_permission(self, source: Any, info: OauthInfo, **kwargs) -> bool:  # type: ignore
+        if not os.environ.get("OAUTH2_ACTIVE"):
+            return True
+
+        service_name = f"/{os.environ.get('SERVICE_NAME') or ''}"
+        path = f"{service_name}/{get_path_as_string(info.path)}/".lower()  # type: ignore
+
+        context = info.context  # type: ignore
+        try:
+            logger.debug("Request headers", headers=info.context.request.headers)  # type: ignore
+            current_user = await context.get_current_user(info.context.request)  # type: ignore
+        except HTTPException:
+            self.message = f"User is not authorized to query or has an invalid access token for path: `{path}`"
+            return False
+
+        opa_decision: bool = await context.get_opa_decision(path, current_user)
+
+        logger.debug("Get opa decision", path=path, opa_decision=opa_decision)
+        if not opa_decision:
+            self.message = f"User is not authorized to query `{path}`"
+
+        return opa_decision
+
+
+class IsAuthenticatedForMutation(BasePermission):
+    message = "User is not authenticated"
+
+    async def has_permission(self, source: Any, info: OauthInfo, **kwargs) -> bool:  # type: ignore
+        mutations_active = os.environ.get("OAUTH2_ACTIVE") and os.environ.get("MUTATIONS_ENABLED")
+        env_ignore_mutation_disabled: list[str] = os.environ.get("ENVIRONMENT_IGNORE_MUTATION_DISABLED") or []  # type: ignore
+        service_name = f"/{os.environ.get('SERVICE_NAME') or ''}"
+
+        if not mutations_active:
+            return os.environ.get("ENVIRONMENT") in env_ignore_mutation_disabled
+
+        path = f"{service_name}/{info.path.key}"  # type: ignore
+        try:
+            current_user = await info.context.get_current_user(info.context.request)  # type: ignore
+        except HTTPException:
+            self.message = f"User is not authorized to query or has an invalid access token for path: `{path}`"
+            return False
+        opa_decision: bool = await info.context.get_opa_decision(path, current_user)  # type: ignore
+
+        logger.debug("Get opa decision", path=path, opa_decision=opa_decision)
+        if not opa_decision:
+            self.message = f"User is not authorized to execute mutation `{path}`"
+
+        return opa_decision
```

Comment:
Should be possible to test these by setting up a strawberry schema, see strawberry-tests for an example https://github.com/strawberry-graphql/strawberry/blob/main/tests/schema/test_permission.py#L32

Reply:
had to do some more tedious stuff to make it work with our authentication classes but added tests for them.